src/inmanta/loader.py

```diff
@@ -213,44 +213,57 @@ def _load_module(self, mod_name: str, hv: str) -> None:
+                if normdir == package_dir:
+                    return
+                pathlib.Path(os.path.join(normdir, "__init__.py")).touch()
+                touch_inits(os.path.dirname(normdir))
+
+            # ensure correct package structure
+            os.makedirs(module_dir, exist_ok=True)
+            touch_inits(os.path.dirname(module_dir))
+            source_file = os.path.join(module_dir, "__init__.py")
+
+            # write the new source
+            with open(source_file, "w+", encoding="utf-8") as fd:
+                fd.write(module_source.source)
+            return True
+        else:
+            LOGGER.debug(
+                "Not deploying code (hv=%s, module=%s) because of cache hit", module_source.hash_value, module_source.name
+            )
+            return False
+
     def deploy_version(self, module_sources: Iterable[ModuleSource]) -> None:
-        to_reload: List[ModuleSource] = []
-
-        for module_source in module_sources:
-            # if the module is new, or update
-            if module_source.name not in self.__modules or module_source.hash_value != self.__modules[module_source.name][0]:
-                LOGGER.info("Deploying code (hv=%s, module=%s)", module_source.hash_value, module_source.name)
-
-                all_modules_dir: str = os.path.join(self.__code_dir, MODULE_DIR)
-                relative_module_path: str = PluginModuleLoader.convert_module_to_relative_path(module_source.name)
-                # Treat all modules as a package for simplicity: module is a dir with source in __init__.py
-                module_dir: str = os.path.join(all_modules_dir, relative_module_path)
-
-                package_dir: str = os.path.normpath(
-                    os.path.join(all_modules_dir, pathlib.PurePath(pathlib.PurePath(relative_module_path).parts[0]))
-                )
-
-                def touch_inits(directory: str) -> None:
-                    """
-                    Make sure __init__.py files exist for this package and all parent packages. Required for compatibility
-                    with pre-2020.4 inmanta clients because they don't necessarily upload the whole package.
-                    """
-                    normdir: str = os.path.normpath(directory)
-                    if normdir == package_dir:
-                        return
-                    pathlib.Path(os.path.join(normdir, "__init__.py")).touch()
-                    touch_inits(os.path.dirname(normdir))
-
-                # ensure correct package structure
-                os.makedirs(module_dir, exist_ok=True)
-                touch_inits(os.path.dirname(module_dir))
-                source_file = os.path.join(module_dir, "__init__.py")
-
-                # write the new source
-                with open(source_file, "w+", encoding="utf-8") as fd:
-                    fd.write(module_source.source)
-
-                to_reload.append(module_source)
+        to_reload: Set[ModuleSource] = set()
```

Comment:
lists replaced by sets to prevent duplicates here

Reply:
This does introduce some nondeterminism but I don't think we really need to worry about that outside of the compiler flow, do we?

Reply:
Also, I don't really see the benefit: we iterate over a set, so each `module_source` should only be added at most once to `to_reload`. Not to say that there is a strong reason to use a list here, either is fine for me.

Reply:
1. the nondeterminism is not a problem here 
2. agreed